janitor/utils.py

```diff
@@ -405,3 +406,219 @@ def _wrapped(x, *args, **kwargs):
+        # flatten structured array into 1d array
+        outcome = np.concatenate(outcome)
+        # sort array
+        outcome.sort(axis=0, order=list(dicts))
+        # create dataframe
+        # structed array already has names,
+        # this gets transferred as column names
+        final = pd.DataFrame.from_records(outcome)
+    return final
+
+
+# this is for dataframes/series
+# this should be the final output if there are lists or lists and dicts
+# returned from check instance
+def _compute_two_dfs(df1: pd.DataFrame, df2: pd.DataFrame):
+    """
+    This computes the cartesian product of two dataframes.
+    Applies to the expand_grid function.
+    Numpy is employed here, to get faster computations,
+    compared to running a many-to-many join with pandas merge.
+    Structured arrays are employed, to preserve data type.
+    """
+    # get lengths of dataframes(number of rows) and swap
+    # essentially we'll pair one dataframe with the other's length:
+    lengths = reversed([ent.index.size for ent in (df1, df2)])
+
+    # grab the maximum string length
+    string_cols = [
+        frame.select_dtypes(include="object").columns for frame in (df1, df2)
+    ]
+
+    # pair max string length with col
+    # will be passed into frame.to_records,
+    # to get dtype in numpy recarray
+    string_cols = [
+        {col: f"<U{frame[col].str.len().max()}" for col in ent}
+        for ent, frame in zip(string_cols, (df1, df2))
+    ]
+
+    # pair length, column data type and dataframe
+    (len_first, col_dtypes, first), (len_last, col_dtypes, last) = list(
+        zip(lengths, string_cols, (df1, df2))
+    )
+
+    # export to numpy as recarray,
+    # ensuring that the column data types are captured
+    # this is particularly relevant to object data type
+    first = first.to_records(column_dtypes=col_dtypes, index=False)
+
+    # tile first with len_first
+    # remember, len_first is the length of the other dataframe
+    first = np.tile(first, (len_first, 1))
+
+    # get a 1d array
+    first = np.concatenate(first)
+
+    # sorting here ensures we get each row of the first
+    # with the entire rows of the other dataframe
+    np.recarray.sort(first, order=first.dtype.names[0])
+
+    # same process as first, except there'll be no sorting
+    last = last.to_records(column_dtypes=col_dtypes, index=False)
+    last = np.tile(last, (len_last, 1))
+    last = np.concatenate(last)
+
+    # merge first and last
+    # and return a dataframe
+    result = rfn.merge_arrays((first, last), flatten=True, asrecarray=True)
+    return pd.DataFrame.from_records(result)
+
+
+def _grid_computation_list(dfs: List):
+    """
+    This computes cartesian product of dataframes in the expand_grid function.
+    This builds on _compute_two_dfs function,
+    by applying it to more than two dataframes.
+    """
+    return functools.reduce(_compute_two_dfs, dfs)
+
+
+def _grid_computation(dfs: List, dicts: Dict):
```

Comment:
Function signature should show return type.

```suggestion
def _grid_computation(dfs: List, dicts: Dict) -> pd.DataFrame:
```